confmap/provider/s3mapprovider/provider.go

```diff
@@ -0,0 +1,138 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package s3mapprovider // import "github.com/open-telemetry/opentelemetry-collector-contrib/confmap/provider/s3mapprovider"
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"go.opentelemetry.io/collector/confmap"
+	"gopkg.in/yaml.v2"
+)
+
+const (
+	schemeName = "s3"
+)
+
+type s3Client interface {
+	GetObject(context.Context, *s3.GetObjectInput, ...func(*s3.Options)) (*s3.GetObjectOutput, error)
+}
+
+type provider struct {
+	client s3Client
+}
+
+// New returns a new confmap.Provider that reads the configuration from a file.
+//
+// This Provider supports "s3" scheme, and can be called with a "uri" that follows:
+//
+//	s3-uri : s3://[BUCKET].s3.[REGION].amazonaws.com/[KEY]
+//
+// One example for s3-uri be like: s3://DOC-EXAMPLE-BUCKET.s3.us-west-2.amazonaws.com/photos/puppy.jpg
+//
+// Examples:
+// `s3://DOC-EXAMPLE-BUCKET.s3.us-west-2.amazonaws.com/photos/puppy.jpg` - (unix, windows)
+func New() confmap.Provider {
+	return &provider{client: nil}
+}
+
+func (fmp *provider) Retrieve(ctx context.Context, uri string, _ confmap.WatcherFunc) (*confmap.Retrieved, error) {
+	if !strings.HasPrefix(uri, schemeName+":") {
+		return nil, fmt.Errorf("%q uri is not supported by %q provider", uri, schemeName)
+	}
+
+	// initialize the s3 client in the first call of Retrieve
+	if fmp.client == nil {
+		cfg, err := config.LoadDefaultConfig(context.Background())
+		if err != nil {
+			fmt.Println("Failed to load configurations to initialize an AWS SDK client, error: %w", err)
```

Comment:
My only concern is that we are using `fmt.Println` here. We can't use the logger, since configuration loading happens before we have a logger, so I don't know if we can do any better. Any thoughts @jpkrohling ?

Reply:
It feels what we really need is to return an error in case the configuration cannot be loaded: there is no point in trying to move on with the creation of a s3 client if the configuration has a problem.

Reply:
Oh, yeah, you are right, I misread the following line. This should just return  an error.

Reply:
Yes, thanks for the suggestions, you are right. Fixed yet!